bundles/core/src/main/java/com/adobe/cq/commerce/demandware/connection/DemandwareClientProviderImpl.java

```diff
@@ -0,0 +1,115 @@
+ ~ You may obtain a copy of the License at
+ ~
+ ~     http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+package com.adobe.cq.commerce.demandware.connection;
+
+import com.adobe.cq.commerce.demandware.DemandwareClient;
+import com.adobe.cq.commerce.demandware.DemandwareClientProvider;
+import com.day.cq.replication.AgentConfig;
+import com.google.common.collect.Maps;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.felix.scr.annotations.Component;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.ReferenceCardinality;
+import org.apache.felix.scr.annotations.ReferencePolicy;
+import org.apache.felix.scr.annotations.Service;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+@Component(label = "Demandware Demandware Client Provider", immediate = true)
+@Service(value = DemandwareClientProvider.class)
+public class DemandwareClientProviderImpl implements DemandwareClientProvider {
+	private static final Logger LOG = LoggerFactory.getLogger(DemandwareClientProviderImpl.class);
+	private static final String DWRE_SCHEME = "demandware://";
+	public static final String INSTANCE_ID_DEFAULT = "default";
+	@Reference(referenceInterface = DemandwareClient.class,
+			bind = "bindDemandwareClient",
+			unbind = "unbindDemandwareClient",
+			cardinality = ReferenceCardinality.MANDATORY_MULTIPLE, // TODO check if works with one
+			policy = ReferencePolicy.DYNAMIC)
+	protected HashMap<String, DemandwareClient> demandwareClients;
+
+	/**
+	 * Returns DemandwareClient service, configured with "default" instance id
+	 * or, if not present, first found DemandwareClient service.
+	 *
+	 * @return
+	 *///TODO remove usages
+	public DemandwareClient getDefaultClient() {
+		Optional<DemandwareClient> client = getClientForSpecificInstance(INSTANCE_ID_DEFAULT);
+		if (client.isPresent()) {
+			return client.get();
+		}
+
+		LOG.debug("No default DemandwareClient configured: return first found service.");
+		Optional<DemandwareClient> clientOpt = demandwareClients.entrySet()
+				.stream()
+				.findFirst()
+				.map(Map.Entry::getValue);
+		if (clientOpt.isPresent()) {
+			return clientOpt.get();
+		}
+
+		LOG.error("Failed to get DemandwareClient - no configuration found.");
+		return null;
+	}
+
+	@Override
+	public Optional<DemandwareClient> getClientForSpecificInstance(final String instanceId) {
+		DemandwareClient demandwareClient = demandwareClients.get(instanceId.replace("/", ""));
```

Comment:
we are defining the names ourselves. where does the "/" come from? and why do we have to remove it?